feat: Add a way to normalize Area and Bar charts + animations continuation

bprusinowski opened the pull request:
- Shared chart variables getters
- Added logic to normalize the Area and Stacked Bar charts (not yet  connected to the UI option)
- Changes the aspect ratios of Pie and Scatter charts to fit on the screen (only in Edit mode)
- Interactive time range filter is now constrained by filters set in the X field
- Interactive time range filter now keeps component iri, which previously was always empty
- Time slider values are now shown in filters list at the top of the charts
- Tooltip message is now displayed when user wants to de-select last color legend item
- Added TemporalOrdinalDimension (and a way to use it with TimeSlider)
- Whiskers are now animated

Rdataflow commented:
@kronmar stay tuned on _ordinal time_ and see https://visualization-tool-purx4zcwb-ixt1.vercel.app/create/new?cube=https://environment.ld.admin.ch/foen/nfi/nfi_ordinal_time/cube/2023-1&flag__timeslider=true (currently WIP, publish is broken)

ptbrowne commented on app/charts/column/columns-stacked-state.tsx:
```diff
@@ -108,22 +108,15 @@ const useColumnsStackedState = (
+        scalesData,
+        (v) => sum(v, (x) => getY(x)),
+        (x) => getSegment(x)
+      ),
```
Here we need to spread because fromEntries does not take an iterator ?

bprusinowski replied:
In fact we don't need to spread, good catch!

ptbrowne commented on app/charts/shared/chart-data-filters.tsx:
```diff
@@ -334,17 +336,16 @@ const DataFilterTemporalDimension = ({
+      timeFormat,
+      formatLocale
+    );
+  }, [options, timeUnit, timeFormat, formatLocale]);
```
It seems you would like the functionality "Add {} to arrow function" from the "JS Refactoring" extension 😁 

<img width="505" alt="image" src="https://github.com/visualize-admin/visualization-tool/assets/465582/c85de33f-aaa1-4dd7-8db1-af9dbde8481a">

<img width="150" alt="image" src="https://github.com/visualize-admin/visualization-tool/assets/465582/1f230308-d553-4db9-9c2d-c3e6a54f2e9a">

bprusinowski replied:
Ooo, definitely! 👀

ptbrowne commented on app/locales/fr/messages.po:
```diff
@@ -333,6 +333,10 @@ msgstr "Tout sélectionner"
 
+#: app/charts/shared/legend-color.tsx
+msgid "controls.filters.interactive.color.min-1-filter"
+msgstr "Au moins un filtre doit être sélectionné."
```
💯

ptbrowne commented on app/charts/chart-config-ui-options.ts:
```diff
@@ -6,7 +6,7 @@ import {
   SortingType,
 } from "@/config-types";
-import { isTemporalDimension } from "@/domain/data";
+import { isTemporalDimension, isTemporalOrdinalDimension } from "@/domain/data";
```
Is it to support the inventory time periods from WSL ?

bprusinowski replied:
Exactly :)

ptbrowne commented on app/configurator/interactive-filters/time-slider.tsx:
```diff
@@ -169,14 +191,49 @@ const Slider = () => {
+    IFState.timeSlider.type,
+    IFState.timeSlider.value,
+    dispatch,
+  ]);
```
Why do we need an useEffet here ? Couldn't the dispatch be done in an event handler ? Or directly in the reducer ?

bprusinowski replied:
Hmmm, I think it can't be done in event handler, as this `useEffect` is also called when timeline is playing (and the `onChange` event only happens when user interacts with the timeline) :(

Or did you mean something else with the reducer? I think we wouldn't have access to timeline in the InteractiveFilters reducer 🤔

ptbrowne commented on app/configurator/interactive-filters/time-slider.tsx:
```diff
@@ -75,17 +75,19 @@ export const TimeSlider = (props: TimeSliderProps) => {
+      new Set(
+        chartState.allData
+          .map((d) => d[component.iri])
+          .filter(Boolean) as string[]
```
I think there is a isTruthy or a truthy helper that can replace the Boolean here, and would help not having to do the `as string[]`.

ptbrowne reviewed (approved):
Thanks Bartosz. The code looks LGTM and I did not see any blocker.

---

General observations

- chart-state
    * Some logic concerning how dimensions are found could be memo-ed ?
    * Maybe we could have an index iri -> dimension/measure to not have to find()
    * Wonder if we could not share code for the dimension hooks useNumerical/useTemporal etc...

    ```
    const useNumericalXVariables = makeVariables({
        dimFilter: isNumericalMeasure,
        field: 'x',
        isOptional: true
        label: getLabelWithUnit
    })
    ```

- I'd like to be able to exercise the changes functionally. It's a bit hard right now since I need to know which charts to exercise, with the right options etc... I think it'd be good to invest some time in developing stories. I think we have the start of this via the mock chart config. Maybe having storybook could help here ?

bprusinowski commented:
Thanks for reviewing @ptbrowne! I addressed the comments in #1105 and created a new issue to add Storybook (#1109).

For the shared logic to create variable getters, I am not sure it would be super easy, as for some variables we use abbreviations, and for other (measures) we don't (also we need custom names for variables, like getX, getY, etc, so I think we would need to unpack, rename and then pass them further)? Let me know if I missed something in this approach!
